src/toxcore/dht/lan_discovery.rs

```diff
@@ -1,20 +1,102 @@
 //! Module for LAN discovery.
 
 use std::iter;
-use std::io::{Error, ErrorKind};
 use std::net::{IpAddr, SocketAddr};
 use std::time::{Duration, Instant};
 
-use futures::{future, stream, Future, Stream};
+use std::fmt;
+use std::error::{Error as StdError};
+use std::io::Error as IoError;
+use std::io::ErrorKind as IoErrorKind;
+
+use failure::{Backtrace, Context, Fail};
+use futures::{stream, Future, Stream};
 use futures::sync::mpsc;
 use get_if_addrs;
 use get_if_addrs::IfAddr;
 use tokio::timer::Interval;
+use tokio::timer::Error as WakeupError;
+use tokio::timer::timeout::Error as TimerError;
 
 use toxcore::crypto_core::*;
 use toxcore::io_tokio::*;
 use toxcore::dht::packet::*;
 
+/// Error that can happen during lan discovery
+#[derive(Debug)]
+pub struct LanDiscoveryError {
+    ctx: Context<LanDiscoveryErrorKind>,
+}
+
+impl LanDiscoveryError {
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &LanDiscoveryErrorKind {
+        self.ctx.get_context()
+    }
+}
+
+impl Fail for LanDiscoveryError {
+    fn cause(&self) -> Option<&Fail> {
+        self.ctx.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.ctx.backtrace()
+    }
+}
+
+impl fmt::Display for LanDiscoveryError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.ctx.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(Debug, Fail)]
+pub enum LanDiscoveryErrorKind {
+    /// Ping wakeup timer error
+    #[fail(display = "Lan discovery wakeup timer error")]
+    Wakeup,
+    /// Send packet(s) error
+    #[fail(display = "Send packet(s) error")]
+    SendTo,
```

Comment:
There shouldn't be such error. It's possible that we can't send a packet due to busy socket and timeout. In this case we don't want to stop node/client.

Reply:
Or at least it shouldn't be returned by `LanDiscovery::run`. Which means we'd probably need separate errors for `LanDiscovery::run` and `LanDiscovery::send`.

Reply:
done